src/smartpeak/source/ui/ChromatogramXICPlotWidget.cpp

```diff
@@ -0,0 +1,95 @@
+// --------------------------------------------------------------------------
+//   SmartPeak -- Fast and Accurate CE-, GC- and LC-MS(/MS) Data Processing
+// --------------------------------------------------------------------------
+// Copyright The SmartPeak Team -- Novo Nordisk Foundation 
+// Center for Biosustainability, Technical University of Denmark 2018-2021.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL ANY OF THE AUTHORS OR THE CONTRIBUTING
+// INSTITUTIONS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Bertrand Boudaud $
+// $Authors: Bertrand Boudaud $
+// --------------------------------------------------------------------------
+
+#include <SmartPeak/ui/ChromatogramXICPlotWidget.h>
+#include <SmartPeak/ui/SpectraMSMSPlotWidget.h>
+#include <implot.h>
+
+namespace SmartPeak
+{
+  void ChromatogramXICPlotWidget::updateData()
+  {
+    // Update marker with sub graph's value
+    if (use_markers_)
+    {
+      auto rt = spectra_msms_plot_widget_->getCurrentRT();
+      marker_position_ = getNearestPoint(rt);
+    }
+
+    std::set<std::string> sample_names = getSelectedSampleNames();
+    std::set<std::string> transitions_names = getSelectedTransitions();
+
+    if ((refresh_needed_) || // data changed
+       ((input_component_names_ != transitions_names) || (input_sample_names_ != sample_names))) // user select different items
```

Comment:
Would it make any benefit to extract out some of these logical functions in the various widgets and encapsulate them into their own methods (e.g., unit testing the logic)?  Similar logic checks in the widgets have caused us numerous problems in the widgets before.

Reply:
I see, ok, i'll improve it.

Reply:
Actually, i don't see too much to improve there. however, I did some factorization since your comment. It's possible to do some unit tests on that, most of these functions are not ui-dependance like this one (updateData). it will be a little bit tedious as it needs some data to check if everything is right. i will try to do some.

Reply:
This can be made into separate user stories if you want and taken up at another time.